src/test_bst.py

```diff
@@ -0,0 +1,192 @@
+
+
+def test_larger_child():
+    """Test that inserting a value greater than head sets the left child of head."""
+    from bst import BST
+    new_bst = BST()
+    new_bst.insert(5)
+    new_bst.insert(9)
+    assert new_bst.head.get_right_child().data == 9
+
+
+def test_insert_same():
+    """Test that a value cannot be added to tree if it already exisits."""
+    from bst import BST
+    new_bst = BST()
+    new_bst.insert(5)
+    with pytest.raises(ValueError):
+        new_bst.insert(5)
+
+
+@pytest.mark.parametrize('value, result', NODES_IN_TREE)
+def test_contains(populated_tree, value, result):
+    """Test that contains method returns correct T/F correctly if node is/isn't in tree."""
+    assert populated_tree.contains(value) == result
+
+
+def test_not_contains(populated_tree):
+    """Test that a value is not in tree."""
+    assert not populated_tree.contains(1)
+
+
+def test_size_empty():
+    """Test that size of empty tree is 0."""
+    from bst import BST
+    new_bst = BST()
+    assert new_bst.size() == 0
+
+
+def test_size_not_empty(populated_tree):
+    """Test that size of populated tree is returned correctly."""
+    assert populated_tree.size() == 7
+
+
+def test_depth(populated_tree):
+    """Test depth of populated tree."""
+    assert populated_tree.depth() == 4
+
+
+def test_balance(populated_tree):
+    """Test balance of left higher tree."""
+    assert populated_tree.balance() == 1
+
+
+def test_balance_empty_tree():
+    """Test balance of empty tree."""
+    from bst import BST
+    new_bst = BST()
+    assert new_bst.balance() == 0
+
+
+def test_balance_1_node():
+    """Test balance of tree of only head."""
+    from bst import BST
+    new_bst = BST()
+    new_bst.insert(10)
+    assert new_bst.balance() == 0
+
+
+def test_right_tree(right_balance_tree):
+    """Test balance of right higher tree."""
+    assert right_balance_tree.balance() == -1
+
+
+def test_preorder(populated_tree):
+    assert [x for x in populated_tree.preorder(populated_tree.head)] == [20, 14, 3, 6, 17, 22, 21]
+
+
+def test_preorder2(right_balance_tree):
+    assert [x for x in right_balance_tree.preorder(right_balance_tree.head)] == [30, 10, 8, 15, 60, 40, 50, 45]
+
+
+def test_inorder(populated_tree):
+    assert [x for x in populated_tree.inorder(populated_tree.head)] == [3, 6, 14, 17, 20, 21, 22]
+
+
+def test_inorder2(right_balance_tree):
+    assert [x for x in right_balance_tree.inorder(right_balance_tree.head)] == [8, 10, 15, 30, 40, 45, 50, 60]
+
+
+def test_post_order(populated_tree):
+    assert [x for x in populated_tree.post_order(populated_tree.head)] == [6, 3, 17, 14, 21, 22, 20]
+
+
+def test_post_order2(right_balance_tree):
+    assert [x for x in right_balance_tree.post_order(right_balance_tree.head)] == [8, 15, 10, 45, 50, 40, 60, 30]
+
+
+def test_breadth(populated_tree):
+    assert [x for x in populated_tree.breadth_first(populated_tree.head)] == [20, 14, 22, 3, 17, 21, 6]
+
+
```

Comment:
Looks like I can cause AttributeErrors by calling traversals on empty trees. This means that:
 - Your code doesn't handle such an event
 - Your tests don't cover such an event